### DISCUSSION TITLE
Just cookie, or also authorization header support?

### DISCUSSION THREAD

**Discussion by askasp** [maxcountryman/axum-login#214] [Q&A] [answered]
I want to use it for my rust backend. However the frontend is both a webapp and a native app. Does it support to extract the user from an access_token in the Authorization header, or just from a cookie?

**Comment by maxcountryman** [comment, marked as answer]
This crate is loosely based on [the Django auth module](https://docs.djangoproject.com/en/5.0/topics/auth/#overview) which in turn is modeled around sessions.

That said, `Credentials` may be anything you like, and it's not difficult to map headers to the authentication logic:

```rust
use headers::Authorization;
use axum_extra::TypedHeader;

async fn auth_header_login(
        mut auth_session: AuthSession,
        TypedHeader(auth_header): TypedHeader<Authorization>,
    ) -> impl IntoResponse {
        let user = match auth_session.authenticate(auth_header).await {
            Ok(Some(user)) => user,
            Ok(None) => return StatusCode::UNAUTHORIZED.into_response(),
            Err(_) => return StatusCode::INTERNAL_SERVER_ERROR.into_response(),
        };

        if auth_session.login(&user).await.is_err() {
            return StatusCode::INTERNAL_SERVER_ERROR.into_response();
        }

        Redirect::to("/").into_response()
    }
}
```

Note that this still uses a session cookie under the covers and so if that e.g. doesn't work in a mobile app context, this probably wouldn't help your use case.

It's worth pointing out that this can also be implemented as a middleware via [`from_fn`](https://docs.rs/axum/latest/axum/middleware/fn.from_fn.html) which obviates the need to rely on a cookie:

```rust
use axum::{
    Router,
    http,
    routing::get,
    response::Response,
    middleware::{self, Next},
    extract::Request,
};
use axum_extra::TypedHeader;
use headers::Authorization;

async fn auth_header_required_middleware(
        mut auth_session: AuthSession,
        TypedHeader(auth_header): TypedHeader<Authorization>,
        request: Request,
        next: Next,
    ) -> impl IntoResponse {
        let user = match auth_session.authenticate(auth_header).await {
            Ok(Some(user)) => user,
            Ok(None) => return StatusCode::UNAUTHORIZED.into_response(),
            Err(_) => return StatusCode::INTERNAL_SERVER_ERROR.into_response(),
        };

        if auth_session.login(&user).await.is_err() {
            return StatusCode::INTERNAL_SERVER_ERROR.into_response();
        }

        next.run(request).await
    }
}

let app = Router::new()
    .route("/api/protected", get(|| async { /* ... */ }))
    .layer(middleware::from_fn(auth_header_required_middleware));
```

**Reply by askasp** [reply]
Thanks again!

**Reply by maxcountryman** [reply]
Happy to help! I think the middleware pattern is easy enough to implement, but if there are ways we can provide better support directly in the crate I'm open to considering that as well.

**Reply by askasp** [reply]
Ill think about it when I implement it. My skills in rust are not very impressive yet, but if I think of a good way ill open a PR
